File: util/create_install_shell_script.pl

#!/usr/bin/env perl
use strict;
use File::Basename;
use File::Compare;
use File::Copy;
use File::Path;
use File::Slurp;
use FindBin;
use Getopt::Long qw(:config no_ignore_case no_auto_abbrev);
use Term::ANSIColor;

use constant FALSE => 0;
use constant TRUE => 1;
use constant DEFAULT_OUTDIR => '/tmp/' . File::Basename::basename($0) . '/' . time();
use constant DEFAULT_VERBOSE => TRUE;

my $help;
my $infile;
my $man;
my $outdir;
my $outfile;
my $verbose;

my $results = GetOptions (
    'help|h'        => \$help,
    'infile=s'      => \$infile, 
    'man|m'         => \$man,
    'outdir=s'      => \$outdir,
    'outfile=s'     => \$outfile,
    'verbose=s'     => \$verbose,
    );

&checkCommandLineArguments();

main();

if ($verbose){
    printGreen(File::Spec->rel2abs($0) . " execution completed");
}

exit(0);

##--------------------------------------------------------
##
##  END OF MAIN -- SUBROUTINES FOLLOW
##
##--------------------------------------------------------

sub checkCommandLineArguments {
   
   	my $fatalCtr = 0;

    if (!defined($infile)){
        printBoldRed("--infile was not specified");
        $fatalCtr++;
    }

    if ($fatalCtr > 0){
    	die "Required command-line arguments were not specified\n";
    }

    if (!defined($verbose)){
        $verbose = DEFAULT_VERBOSE;
        printYellow("--verbose was not specified and therefore was set to default '$verbose'");
    }

    if (!defined($outdir)){
        $outdir = DEFAULT_OUTDIR;
        printYellow("--outdir was not specified and therefore was set to default '$outdir'");        
    }

    if (!-e $outdir){
        mkpath($outdir) || die "Could not create output directory '$outdir' : $!";
        print "Created output directory '$outdir'\n";
    }

    if (!defined($outfile)){
        my $basename = File::Basename::basename($infile);
        $basename =~ s/\.txt$//;
        $outfile = $outdir . '/install_'. $basename . '.sh';
        printYellow("--outfile was not specified and therefore was set to default '$outfile'");
    }
}

sub printBoldRed {

    my ($msg) = @_;
    print color 'bold red';
    print $msg . "\n";
    print color 'reset';
}

sub printGreen {

    my ($msg) = @_;
    print color 'green';
    print $msg . "\n";
    print color 'reset';
}

sub printYellow {

    my ($msg) = @_;
    print color 'yellow';
    print $msg . "\n";
    print color 'reset';
}

sub main {

    if (!-e $infile){
        die "infile '$infile' does not exist";
    }

    if (!-f $infile){
        die "infile '$infile' is not a regular file";
    }

    if (!-s $infile){
        die "infile '$infile' does not have any content";
    }

    open (OUTFILE, ">$outfile") || die "Could not open file '$outfile' in write mode : $!";
    
    print OUTFILE "#!/bin/bash\n\n";

    my @content = read_file($infile);
    
    chomp @content;
    
    for my $line (@content){
    
        if ($line =~ /^Reference:\s*(.+)/){
            print OUTFILE "echo \"Reference: $1\"\n\n";
        }
        else {
            print OUTFILE "echo \"About to execute: $line\"\n";
            print OUTFILE $line . "\n\n";
        }
    }

    close OUTFILE;
    print "Wrote output file '$outfile'\n";
}

File: util/create_install_shell_script.py
import os
import sys

DEFAULT_OUTFILE = os.path.join('/tmp', os.path.basename(__file__) + '.sh')
outfile = None
infile = sys.argv[1]

if len(sys.argv) == 3:
    outfile = sys.argv[2]

if infile is None:
    print("Usage: python __file__ [input text file with commands] [output bash shell script file]")
    sys.exit(1)

if not os.path.exists(infile):
    raise Exception(f"File '{infile}' does not exist")

if outfile is None:
    basename = os.path.basename(infile)
    if basename.endswith('.txt'):
        basename = basename.replace('.txt', '')
    outfile = os.path.join('/tmp', basename + '.sh')
    print(f"outfile was not specified and therefore was set to '{outfile}'")

with open(outfile, 'wt') as of:
    of.write("#!/bin/sh\n\n")
    with open(infile, 'r') as f:
        for line in f:
            line = line.strip()
            if 'reference' in line.lower():
                of.write(f"{line}\n")
            else:
                of.write(f'echo "Will attempt to execute: {line}"\n')
                of.write(f"{line}\n\n")

    print(f"Wrote shell script '{outfile}'")